src/core/ddsi/src/ddsi_deadline.c

```diff
@@ -105,8 +107,23 @@ void ddsi_deadline_renew_instance_real (struct ddsi_deadline_adm *deadline_adm,
      will still be triggered, but has no effect on this instance because in
      the callback the deadline (which will be the updated value) will be
      checked for expiry */
+  ddsrt_mtime_t updatetime = ddsrt_time_monotonic();
+  int64_t deadlines_missed = (updatetime.v - elem->t_last_trigger.v)/deadline_adm->dur;
+  if (deadlines_missed > 0)
+  {
+    if ((deadlines_missed + elem->deadlines_missed) > UINT32_MAX)
+    {
+      assert(false);
+      elem->deadlines_missed = UINT32_MAX;
+    }
+    else
+    {
+      elem->deadlines_missed += (uint32_t)deadlines_missed;
+    }
+  }
+
   ddsrt_circlist_remove(&deadline_adm->list, &elem->e);
-  elem->t_deadline = ddsrt_time_monotonic();
-  elem->t_deadline.v += deadline_adm->dur;
+  elem->t_last_trigger = updatetime;
```

Comment:
So I think this `t_last_trigger` would be better as a `t_last_update` and leave it unchanged in case of a deadline miss. Because then the list is actually ordered on `t_last_update`, and so still maintains the invariant that is a prefix of entries that have a missed deadline, followed only by ones that have not missed it yet.

And then it probably is no longer useful to maintain `deadlines_missed` in the elements, because you can simply compute it in the handler.